bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/client/MqttAsyncClientWrapper.java

```diff
@@ -72,23 +72,29 @@
      * @return a CompletableFuture (exceptionally on fail)
      */
     public CompletableFuture<Boolean> subscribe(String topic, int qos, ClientCallback clientCallback) {
-        subscribedTopics.add(topic);
-        return internalSubscribe(topic, qos, clientCallback).thenApply(s -> {
-            logger.trace("successfully subscribed {} to topic {}", this, topic);
-            return true;
-        });
+        boolean needsSubscription = subscriptions.add(clientCallback);
+        if (needsSubscription) {
+            logger.trace("Trying to subscribe {} to topic {}", this, topic);
+            return internalSubscribe(topic, qos, clientCallback).thenApply(s -> {
+                logger.trace("Successfully subscribed {} to topic {}", this, topic);
+                return true;
+            });
+        } else {
+            logger.trace("{} already subscribed to topic {}", this, topic);
+            return CompletableFuture.completedFuture(true);
```

Comment:
We could use the result value of the future to indicate, that a real subscription was not needed.
Then we could still do the debug messages in the connection.
Not sure, id that will give us much benefit.

Reply:
I don't think this gives much benefit and currently either true or exceptionally is expected.